src/MeshlessFV/MfvCommon.cpp

```diff
@@ -259,142 +288,59 @@ void MfvCommon<ndim, kernelclass>::ComputePsiFactors
-  int var;                             // Particle state vector summation variable
-  FLOAT draux[ndim];                   // Relative position vector
-  FLOAT drsqd;                         // Distance squared
-  FLOAT dv[ndim];                      // Relative velocity vector
-  FLOAT dvdr;                          // Dot product of dv and dr
-  FLOAT psitilda[ndim];                // Normalised gradient psi factor
-  const FLOAT invhsqd = part.invh*part.invh;   // Local copy of 1/h^2
-
-
-  // Initialise/zero all variables to be updated in this routine
-  part.vsig_max = (FLOAT) 0.0;
-  for (k=0; k<ndim; k++) part.vreg[k] = (FLOAT) 0.0;
-  for (k=0; k<ndim; k++) {
-    for (var=0; var<nvar; var++) {
-      part.grad[var][k] = (FLOAT) 0.0;
-    }
-  }
-
-  // Loop over all potential neighbours in the list
-  //-----------------------------------------------------------------------------------------------
-  for (jj=0; jj<Nneib; jj++) {
-    j = neiblist[jj];
-
-    for (k=0; k<ndim; k++) draux[k] = neibpart[j].r[k] - part.r[k];
     for (k=0; k<ndim; k++) dv[k] = neibpart[j].v[k] - part.v[k];
-    dvdr = DotProduct(dv, draux, ndim);
     drsqd = DotProduct(draux, draux, ndim);
+    dvdr = DotProduct(dv, draux, ndim);
+
+    double w = part.hfactor*kern.w0_s2(drsqd*invhsqd)/part.ndens;
 
-    // Calculate psitilda values
-    for (k=0; k<ndim; k++) psitilda[k] = (FLOAT) 0.0;
     for (k=0; k<ndim; k++) {
       for (int kk=0; kk<ndim; kk++) {
-        psitilda[k] += part.B[k][kk]*draux[kk]*part.hfactor*kern.w0_s2(drsqd*invhsqd)/part.ndens;
+        E[k][kk] += draux[k]*draux[kk]*w ;
       }
     }
 
-    // Calculate contribution to gradient from neighbour
-    for (var=0; var<nvar; var++) {
+    // Compute the first part of gradient, we do the matrix mult later, when part.B is available.
+    for (var=0; var <nvar; var++) {
       for (k=0; k<ndim; k++) {
-        part.grad[var][k] += (neibpart[j].Wprim[var] - part.Wprim[var])*psitilda[k];
+        grad_tmp[var][k] += draux[k]*(neibpart[j].Wprim[var] - part.Wprim[var])*w ;
       }
     }
 
     // Calculate maximum signal velocity
     part.vsig_max = max(part.vsig_max, part.sound + neibpart[j].sound -
-                                       min((FLOAT) 0.0, dvdr/(sqrtf(drsqd) + small_number)));
+        min((FLOAT) 0.0, dvdr/(sqrtf(drsqd) + small_number)));
     part.levelneib = max(part.levelneib, neibpart[j].level) ;
-
-    for (k=0; k<ndim; k++) part.vreg[k] -= draux[k]*kern.w0_s2(drsqd*invhsqd);
-
   }
   //-----------------------------------------------------------------------------------------------
 
-  for (k=0; k<ndim; k++) part.vreg[k] *= part.invh*part.sound;  //pow(part.invh, ndim);
 
+  // Invert the matrix (depending on dimensionality)
+  InvertMatrix<ndim>(E,part.B) ;
 
+  // Complete the calculation of the gradients:
+  for (var=0; var<nvar; var++) {
+    for (k=0; k<ndim; k++) {
+      part.grad[var][k] = DotProduct(part.B[k], grad_tmp[var], ndim) ;
```

Comment:
It works by changing the order of the sum. By avoiding all the extra kernel evaluations it can save a few per cent at least.